util/signing/signer/src/lib.rs

```diff
@@ -0,0 +1,124 @@
+pub mod cryptography;
+
+/// A collection of bytes.
+#[derive(Debug, Clone)]
+pub struct Bytes(pub Vec<u8>);
+
+/// A signature.
+#[derive(Debug, Clone)]
+pub struct Signature(pub Bytes);
+
+/// A public key.
+#[derive(Debug, Clone)]
+pub struct PublicKey(pub Bytes);
+
+/// Version of a key.
+/// Default mean the current key.
+#[derive(Debug, Clone, Default)]
+pub struct KeyVersion(pub String);
+
+/// Id that identify a Key.
+#[derive(Debug, Clone)]
+pub struct KeyId(pub String);
+
+/// Errors thrown by Signer
+#[derive(Debug, thiserror::Error)]
+pub enum SignerError {
+	#[error("Error during signing : {0}")]
+	Sign(String),
+	#[error("Error during public key retrieval : {0}")]
+	PublicKey(String),
+	#[error("Error can't decode provided hex data : {0}")]
+	Hex(String),
+	#[error("Signature not found.")]
+	SignatureNotFound,
+	#[error("public key not found.")]
+	PublicKeyNotFound,
+}
+
+#[async_trait::async_trait]
+pub trait SignerOperations<C: cryptography::Curve> {
+	/// Signs some bytes.
+	async fn sign(&self, message: Bytes) -> Result<Signature, SignerError>;
+
+	/// Gets the public key.
+	async fn public_key(&self) -> Result<PublicKey, SignerError>;
+}
+
+pub struct Signer<O, C>
+where
+	O: SignerOperations<C>,
+	C: cryptography::Curve,
+{
+	operations: O,
+	_curve_marker: std::marker::PhantomData<C>,
+}
+
+/// Signer wraps an implementation of [SignerOperations] and provides a simple API for signing and getting the public key.
+impl<O, C> Signer<O, C>
+where
+	O: SignerOperations<C>,
+	C: cryptography::Curve,
+{
+	pub fn new(operations: O) -> Self {
+		Self { operations, _curve_marker: std::marker::PhantomData }
+	}
+
+	/// Signs some bytes.
+	pub async fn sign(&self, message: Bytes) -> Result<Signature, SignerError> {
+		self.operations.sign(message).await
+	}
+
+	/// Gets the public key.
+	pub async fn public_key(&self) -> Result<PublicKey, SignerError> {
```

Comment:
This API does not take versioned keys into account. The implementation then must assume or configure a particular derivation path, e.g. that Vault will use key ID "maptos" with version 4. If the key is later rotated, the configuration would have to be updated for the different version.

An alternative is to return the derivation string alongside the signature from the `sign` operation, and then use it as a parameter in `public_key`. But this makes the abstract API cumbersome, as the application does not otherwise have a good reason to manage versioned keys.

@musitdev I think I agree with this simplification on the API level. Let the key version be a configuration parameter for the implementation.

Reply:
Yes, the version can be hidden except if we want to be able to rotate the key while the application is running. In this case, the application should have a way to know that the key has changed.

Reply:
Key rotations I would place under a different trait. See https://github.com/movementlabsxyz/sre/tree/main/secure-signing/signing-operations/key-rotation

Reply:
I agree about rotations needing a different trait, as long as any application we're developing requires abstract key rotation at all.

For setup crates and similar supporting functionality, we can always go for the specific provider APIs directly, because setup would likely need details specific to the provider anyway.